src/MongoDB.Driver.Core/Core/Operations/RetryabilityHelper.cs

```diff
@@ -0,0 +1,98 @@
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+
+namespace MongoDB.Driver.Core.Operations
+{
+    internal static class RetryabilityHelper
+    {
+        // private static fields
+        private static readonly HashSet<Type> __resumableChangeStreamExceptions;
+        private static readonly HashSet<ServerErrorCode> __resumableChangeStreamErrorCodes;
+        private static readonly HashSet<Type> __retryableWriteExceptions;
+        private static readonly HashSet<ServerErrorCode> __retryableWriteErrorCodes;
+
+        // static constructor
+        static RetryabilityHelper()
+        {
+            var resumableAndRetryableExceptions = new HashSet<Type>()
+            {
+                typeof(MongoConnectionException),
+                typeof(MongoNotPrimaryException)
+            };
+
+            __resumableChangeStreamExceptions = new HashSet<Type>(resumableAndRetryableExceptions)
+            {
+                typeof(MongoCursorNotFoundException)
+            };
+
+            __retryableWriteExceptions = new HashSet<Type>(resumableAndRetryableExceptions)
+            {
+                typeof(MongoNodeIsRecoveringException)
+            };
+
+            var resumableAndRetryableErrorCodes = new HashSet<ServerErrorCode>
+            {
+                ServerErrorCode.HostNotFound,
+                ServerErrorCode.HostUnreachable,
+                ServerErrorCode.InterruptedAtShutdown,
+                ServerErrorCode.InterruptedDueToReplStateChange,
+                ServerErrorCode.NetworkTimeout,
+                ServerErrorCode.NotMaster,
+                ServerErrorCode.NotMasterNoSlaveOk,
+                ServerErrorCode.NotMasterOrSecondary,
+                ServerErrorCode.PrimarySteppedDown,
+                ServerErrorCode.ShutdownInProgress,
+                ServerErrorCode.SocketException
+            };
+
+            __resumableChangeStreamErrorCodes = new HashSet<ServerErrorCode>(resumableAndRetryableErrorCodes)
+            {
+            };
+
+            __retryableWriteErrorCodes = new HashSet<ServerErrorCode>(resumableAndRetryableErrorCodes)
+            {
+                ServerErrorCode.WriteConcernFailed
+            };
+        }
+
+        // public static methods
+        public static bool IsResumableChangeStreamException(Exception exception)
+        {
+            return IsMatch(exception, __resumableChangeStreamExceptions, __resumableChangeStreamErrorCodes);
+        }
+
+        public static bool IsRetryableWriteException(Exception exception)
+        {
+            return IsMatch(exception, __retryableWriteExceptions, __retryableWriteErrorCodes);
+        }
+
+        // private static methods
+        private static bool IsMatch(Exception exception, HashSet<Type> matchingExceptions, HashSet<ServerErrorCode> matchingErrorCodes)
```

Comment:
The two methods (IsResumableChangeStreamException and IsRetryableWriteException) share the same implementation but use different sets of matching exceptions and error codes.

Reply:
Nice.